File: src/python/extra/heuristic_detection.py
# Heuristic-based AI Bot Detection
# Advanced detection methods beyond simple user-agent matching

from typing import Dict, List, Optional, Set
from dataclasses import dataclass
from datetime import datetime, timedelta
import time

@dataclass
class HeuristicDetectionConfig:
    """Configuration for heuristic-based bot detection"""
    max_requests_per_minute: int = 60
    suspicious_user_agent_words: List[str] = None
    required_headers: List[str] = None
    min_accept_language_length: int = 5
    enable_robots_check: bool = True
    enable_asset_request_check: bool = True
    
    def __post_init__(self):
        if self.suspicious_user_agent_words is None:
            self.suspicious_user_agent_words = [
                'bot', 'crawler', 'scraper', 'spider', 'fetch',
                'harvest', 'extract', 'scan', 'monitor', 'check'
            ]
        if self.required_headers is None:
            self.required_headers = ['accept', 'accept-language', 'accept-encoding']

@dataclass
class RequestContext:
    """Context information for a request used in heuristic analysis"""
    user_agent: str
    headers: Dict[str, str]
    path: str
    ip: str
    timestamp: datetime
    is_html_request: bool
    requested_assets: List[str] = None
    
    def __post_init__(self):
        if self.requested_assets is None:
            self.requested_assets = []

class IPRequestTracker:
    """Tracks requests from individual IP addresses"""
    
    def __init__(self):
        # TODO: Implement IP-based request tracking data structures
        pass

def is_ai_crawler_by_heuristics(ctx: RequestContext, config: HeuristicDetectionConfig) -> bool:
    """
    Main heuristic detection function that combines multiple detection methods
    
    Args:
        ctx: Request context containing all relevant information
        config: Configuration for heuristic detection
        
    Returns:
        True if the request appears to be from an AI crawler
    """
    # TODO: Implement heuristic-based AI crawler detection
    # This function should combine multiple heuristics for better accuracy
    return False

def has_suspicious_user_agent(user_agent: str, suspicious_words: List[str]) -> bool:
    """
    Check for suspicious User-Agent patterns
    
    TODO: Implement checks for:
    - Generic names like "bot", "crawler", "scraper"
    - Missing version numbers
    - Unusual browser/OS combinations
    - Too short or too long user agents
    - Non-standard formatting
    """
    return False

def has_suspicious_headers(headers: Dict[str, str], required_headers: List[str]) -> bool:
    """
    Analyze request headers for bot-like behavior
    
    TODO: Implement analysis for:
    - Missing Accept header or unusual Accept values
    - Missing or suspicious Accept-Language
    - Missing DNT (Do Not Track) header
    - Missing or unusual Sec-Fetch-* headers
    - Inconsistent header combinations
    - Headers that don't match claimed User-Agent
    """
    return False

def has_high_request_rate(ip: str, tracker: IPRequestTracker, max_per_minute: int) -> bool:
    """
    Check if IP has unusually high request rate
    
    TODO: Implement request rate analysis:
    - Track requests per IP over time windows
    - Compare against normal user patterns
    - Account for legitimate high-traffic scenarios
    """
    return False

def accesses_unlinked_content(path: str, known_linked_paths: Set[str]) -> bool:
    """
    Check if request targets unlinked/hidden content
    
    TODO: Implement detection for:
    - Admin panels, hidden directories
    - Direct file access without referrer
    - Systematic directory traversal patterns
    - Access to files typically not linked (sitemap.xml, etc.)
    """
    return False

def ignores_robots_txt(path: str, user_agent: str, robots_rules: Dict[str, List[str]]) -> bool:
    """
    Check if the requester ignores robots.txt rules
    
    TODO: Implement robots.txt compliance checking:
    - Verify if path is disallowed for this user agent
    - Track patterns of robots.txt violations
    - Consider crawl-delay violations
    """
    return False

def missing_asset_requests(ctx: RequestContext, tracker: IPRequestTracker) -> bool:
    """
    Detect if HTML requests aren't followed by asset requests
    
    TODO: Implement asset request pattern analysis:
    - Track if HTML requests are followed by CSS/JS/image requests
    - Consider normal browser behavior patterns
    - Account for cached resources and repeat visits
    - Analyze timing patterns between requests
    """
    return False

def analyze_behavior_patterns(requests: List[RequestContext]) -> bool:
    """
    Perform comprehensive behavioral analysis
    
    TODO: Implement analysis for:
    - Request timing patterns (too regular/irregular)
    - Navigation patterns (depth-first vs breadth-first)
    - Response to different content types
    - Session duration and page view patterns
    - Interaction with dynamic content
    """
    return False

def create_ip_request_tracker() -> IPRequestTracker:
    """Create a new IP request tracker"""
    # TODO: Initialize IP tracking data structures
    return IPRequestTracker()

def get_default_heuristic_config() -> HeuristicDetectionConfig:
    """Get default configuration for heuristic detection"""
    return HeuristicDetectionConfig()

File: src/js/extra/heuristic_detection.js
// Heuristic-based AI Bot Detection
// Advanced detection methods beyond simple user-agent matching

class HeuristicDetectionConfig {
    constructor({
        maxRequestsPerMinute = 60,
        suspiciousUserAgentWords = [
            'bot', 'crawler', 'scraper', 'spider', 'fetch',
            'harvest', 'extract', 'scan', 'monitor', 'check'
        ],
        requiredHeaders = ['accept', 'accept-language', 'accept-encoding'],
        minAcceptLanguageLength = 5,
        enableRobotsCheck = true,
        enableAssetRequestCheck = true
    } = {}) {
        this.maxRequestsPerMinute = maxRequestsPerMinute;
        this.suspiciousUserAgentWords = suspiciousUserAgentWords;
        this.requiredHeaders = requiredHeaders;
        this.minAcceptLanguageLength = minAcceptLanguageLength;
        this.enableRobotsCheck = enableRobotsCheck;
        this.enableAssetRequestCheck = enableAssetRequestCheck;
    }
}

class RequestContext {
    constructor(userAgent, headers, path, ip, timestamp, isHTMLRequest, requestedAssets = []) {
        this.userAgent = userAgent;
        this.headers = headers;
        this.path = path;
        this.ip = ip;
        this.timestamp = timestamp;
        this.isHTMLRequest = isHTMLRequest;
        this.requestedAssets = requestedAssets;
    }
}

class IPRequestTracker {
    constructor() {
        // TODO: Implement IP-based request tracking data structures
    }
}

// Main heuristic detection function
function isAICrawlerByHeuristics(ctx, config) {
    // TODO: Implement heuristic-based AI crawler detection
    // This function should combine multiple heuristics for better accuracy
    return false;
}

// Check for suspicious User-Agent patterns
function hasSuspiciousUserAgent(userAgent, suspiciousWords) {
    // TODO: Check for suspicious User-Agent patterns:
    // - Generic names like "bot", "crawler", "scraper"
    // - Missing version numbers
    // - Unusual browser/OS combinations
    // - Too short or too long user agents
    // - Non-standard formatting
    return false;
}

// Analyze request headers for bot-like behavior
function hasSuspiciousHeaders(headers, requiredHeaders) {
    // TODO: Analyze headers for suspicious patterns:
    // - Missing Accept header or unusual Accept values
    // - Missing or suspicious Accept-Language
    // - Missing DNT (Do Not Track) header
    // - Missing or unusual Sec-Fetch-* headers
    // - Inconsistent header combinations
    // - Headers that don't match claimed User-Agent
    return false;
}

// Check if IP has unusually high request rate
function hasHighRequestRate(ip, tracker, maxPerMinute) {
    // TODO: Implement request rate analysis:
    // - Track requests per IP over time windows
    // - Compare against normal user patterns
    // - Account for legitimate high-traffic scenarios
    return false;
}

// Check if request targets unlinked/hidden content
function accessesUnlinkedContent(path, knownLinkedPaths) {
    // TODO: Detect access to unlinked content:
    // - Admin panels, hidden directories
    // - Direct file access without referrer
    // - Systematic directory traversal patterns
    // - Access to files typically not linked (sitemap.xml, etc.)
    return false;
}

// Check if the requester ignores robots.txt rules
function ignoresRobotsTxt(path, userAgent, robotsRules) {
    // TODO: Check robots.txt compliance:
    // - Verify if path is disallowed for this user agent
    // - Track patterns of robots.txt violations
    // - Consider crawl-delay violations
    return false;
}

// Detect if HTML requests aren't followed by asset requests
function missingAssetRequests(ctx, tracker) {
    // TODO: Analyze asset request patterns:
    // - Track if HTML requests are followed by CSS/JS/image requests
    // - Consider normal browser behavior patterns
    // - Account for cached resources and repeat visits
    // - Analyze timing patterns between requests
    return false;
}

// Perform comprehensive behavioral analysis
function analyzeBehaviorPatterns(requests) {
    // TODO: Analyze overall behavior patterns:
    // - Request timing patterns (too regular/irregular)
    // - Navigation patterns (depth-first vs breadth-first)
    // - Response to different content types
    // - Session duration and page view patterns
    // - Interaction with dynamic content
    return false;
}

// Create a new IP request tracker
function createIPRequestTracker() {
    // TODO: Initialize IP tracking data structures
    return new IPRequestTracker();
}

// Get default configuration for heuristic detection
function getDefaultHeuristicConfig() {
    return new HeuristicDetectionConfig();
}

export {
    HeuristicDetectionConfig,
    RequestContext,
    IPRequestTracker,
    isAICrawlerByHeuristics,
    hasSuspiciousUserAgent,
    hasSuspiciousHeaders,
    hasHighRequestRate,
    accessesUnlinkedContent,
    ignoresRobotsTxt,
    missingAssetRequests,
    analyzeBehaviorPatterns,
    createIPRequestTracker,
    getDefaultHeuristicConfig
};
